packages/controller-utils/src/util.ts

```diff
@@ -257,14 +257,22 @@ export async function safelyExecuteWithTimeout<Result>(
  * @returns A 0x-prefixed hexidecimal checksummed address, if address is valid. Otherwise original input 0x-prefixe, if address is valid. Otherwise original input 0x-prefixed.
  */
 export function toChecksumHexAddress(address: string) {
```

Comment:
I agree with @legobeat. `| null | undefined` should be part of the `address` param's type signature. 

Especially with `exactOptionalPropertyTypes` enabled, the only clear use case for `| undefined` over `?:` is when `undefined` is explicitly being assigned or passed in, not just an uninstantiated or omitted value. And that's exactly the case here.

I don't think overloading is necessary, especially since type signatures for overloaded functions are not inferred properly by TypeScript anyway.
- See https://github.com/microsoft/TypeScript/issues/57351
- We see this with the `getNetworkClientById` method in `NetworkController`. Its return type is always inferred as `AutoManagedNetworkClient<CustomNetworkClientConfiguration>`.

We also don't need to narrow `address` beyond `string`, so there might not be a strong need to make this generic. 

I think we could do this the simple way:

```suggestion
 * @returns A 0x-prefixed hexidecimal checksummed address, if address is valid. Otherwise original input 0x-prefixe, if address is valid. Otherwise original input 0x-prefixed.
 */
export function toChecksumHexAddress(address: string | null | undefined) {
```

Reply:
Hmm. I see the point being made here, but the problem with changing the argument type is that this also changes the return type. So I think we would need overloads in order to communicate that:

1. When `address` is not a string, the return type is the same
2. When `address` is a string, the return type is a string (and really, it's a `Hex`)

Otherwise, if the argument type is `unknown`, then the return type is inferred to be `unknown`. This causes problems with how it's used in e.g. AddressBookController (and any other place), and the resulting address would need to be checked for whether it is really a hex string. It seems like we should avoid this.

Reply:
@mikesposito That's a great point. @mcmire In that case, using overloads and generics the way @legobeat originally suggested seems like the best way (with nullable types not handled explicitly).

Here's the overloaded function signature:

```ts
/** Overload signatures */
export function toChecksumHexAddress(address: string): `0x${string}`;
export function toChecksumHexAddress<T>(address: T): T;
/** Implementation signature */
export function toChecksumHexAddress(address: unknown): unknown {
  ...
}
```
And here's a [Playground link](https://www.typescriptlang.org/play?#code/PQKhAIHkDcFMCcA2B7AhgE3AZwJYHMA7VAFwFd5YtwRgAoWADwAdl5jwAzUggY2J2QFwxZAGEAFrB4BrLKQC2ACUYBBdOgpYsACgwbKWAFzZi8HATwBKYwAMADAwAkAbyynzeAL42A3PWas7Fy8-ILCYpIyckqq6ppYADwAKgB8unEGxknW4El+oBAAkvJMiLDysATEJAJCuIQk5LDUdIwsbJzcfLXhElKyCsoMavpa6aNG4NzSBMgA7gQ507ML4M604JvgFGTwBH6etLQF4IUEHAhUPKwUfODIcEhomPVEu80crODmTKTsxABPJjNGhHMrsHBYABypEQiGMBFhiFo4O+WAAyu4LMY3GYLEcChsIJi8XhvgRfsQiXRjsBOiEeiI+lFBrEJuN4jisVZbA4XLiPN5wNoANQARnujxQGEstEhMLh4AAvL1IgMYsMMmMAOTayw+cDAOlJIHNbX2JyubnebVo8CzdioLT4IgAIzK4WEpvA2sRcO1ADptAAmADMweDstp9O6YSZauiQxG8Q5mRMpJyFv51pswvFkoQ0vQsshJI8ytV-UTbJT2ocABZYGKOBwAOwRgCcrvUqA4dlQrdD6A4YoAHKHUKPR+gOzwJ3Z6zweB3WwBWWB2PUEsBE8BQwQAWgFFnJlOpBLpwVjQnjVdZmomCVMpFgaT0nOE8BfOWfzVFEoeQtnhLaEkQrW8WQ1ZMDG0X99UNY1vW1X9bUhe1kEdZ0GndZoRC9YEfT9RBAxDcNIwvGNQhvCI7ygrVEm4dBYA4cxYHQN96OMRjmNY4suIIJiWIINi8wAqVgLlDFuXAmjIKTejtG4oS2Pgo1ciQpTeNQqgHXAJ1XlQHDPUBAjtWPPASLDCMozUq8qMrOSawMBIiI4iYESRHIiNEgsnhlSSFUQGTmXVeT2SI-VaCAA) showing that the correct overload signature (and return type) is inferred for string vs. non-string input.

Reply:
My overloaded version (https://github.com/MetaMask/core/pull/4046#discussion_r1523662851) runs into this error:

```ts
packages/preferences-controller/src/PreferencesController.ts:261:24 - error TS2538: Type 'unknown' cannot be used as an index type.

261         if (identities[address]) {
                           ~~~~~~~

packages/preferences-controller/src/PreferencesController.ts:266:20 - error TS2538: Type 'unknown' cannot be used as an index type.

266         identities[address] = {
                       ~~~~~~~
```

This is fixed with the following:

```diff
diff --git a/packages/preferences-controller/src/PreferencesController.ts b/packages/preferences-controller/src/PreferencesController.ts
index 2bf761950..b7451b2c2 100644
--- a/packages/preferences-controller/src/PreferencesController.ts
+++ b/packages/preferences-controller/src/PreferencesController.ts
@@ -254,7 +254,9 @@ export class PreferencesController extends BaseController<
    * @param addresses - List of addresses to use to generate new identities.
    */
   addIdentities(addresses: string[]) {
-    const checksummedAddresses = addresses.map(toChecksumHexAddress);
+    const checksummedAddresses = addresses.map((address) =>
+      toChecksumHexAddress(address),
+    );
     this.update((state) => {
       const { identities } = state;
       for (const address of checksummedAddresses) {

```

This is interesting. I didn't know passing a callback into `map` can interfere with type inference like this.